x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/ChoroplethMap/index.tsx

```diff
@@ -0,0 +1,278 @@
+  const getValueScale = useCallback(
+    (value: number) => (value - min) / (max - min),
+    [max, min]
+  );
+
+  const controlScrollZoomOnWheel = useCallback((event: WheelEvent) => {
+    if (event.ctrlKey || event.metaKey) {
+      event.preventDefault();
+    } else {
+      event.stopPropagation();
+    }
+  }, []);
+
+  // side effect creates a new mouseover handler referencing new component state
+  // and replaces the old one stored in `updateTooltipStateOnMousemoveRef`
+  useEffect(() => {
+    const updateTooltipStateOnMousemove = (
+      event: mapboxgl.MapMouseEvent & mapboxgl.EventData
+    ) => {
+      const isMapQueryable =
+        map &&
+        popupRef.current &&
+        data.length &&
+        map.getLayer(CHOROPLETH_LAYER_ID);
+
+      if (!isMapQueryable) {
+        return;
+      }
+      (popupRef.current as Popup).setLngLat(event.lngLat);
+      const hoverFeatures = (map as Map).queryRenderedFeatures(event.point, {
+        layers: [CHOROPLETH_LAYER_ID]
+      });
+
+      if (!hoverFeatures[0]) {
+        return setTooltipState(null);
+      }
+
+      const geojsonProperties = hoverFeatures[0].properties as NonNullable<
+        GeoJsonProperties
+      >;
+
+      if (tooltipState && tooltipState.name === geojsonProperties.name) {
+        return;
+      }
+
+      const item = data.find(
+        ({ key }) =>
+          geojsonProperties && key === geojsonProperties[GEOJSON_KEY_PROPERTY]
+      );
+      if (item)
+        setTooltipState({
+          name: geojsonProperties.name,
+          value: item.value,
+          docCount: item.docCount
+        });
+    };
+    updateTooltipStateOnMousemoveRef.current = updateTooltipStateOnMousemove;
+  }, [map, data, tooltipState]);
+
+  const updateTooltipStateOnMousemoveRef = useRef(
+    (event: mapboxgl.MapMouseEvent & mapboxgl.EventData) => {}
+  );
+
+  const updateTooltipStateOnMouseout = useCallback(
+    (event: mapboxgl.MapMouseEvent & mapboxgl.EventData) => {
+      enableScrollZoom.current = false;
+      setTooltipState(null);
+    },
+    []
+  );
+
+  // initialization side effect, only runs once
+  useEffect(() => {
+    if (containerRef.current === null) {
+      return;
+    }
+
+    // set up Map object
+    const mapboxMap = new Map({
+      attributionControl: false,
+      container: containerRef.current,
+      dragRotate: false,
+      touchZoomRotate: false,
+      zoom: 0.85,
+      center: { lng: 0, lat: 30 },
+      style: MAPBOX_STYLE
+    });
+
+    mapboxMap.addControl(
+      new NavigationControl({ showCompass: false }),
+      'top-left'
+    );
+
+    // set up Popup object
+    popupRef.current = new Popup({
+      closeButton: false,
+      closeOnClick: false
+    });
+
+    // always use the current handler which changes with component state
+    mapboxMap.on('mousemove', (...args) =>
+      updateTooltipStateOnMousemoveRef.current(...args)
+    );
+    mapboxMap.on('mouseout', updateTooltipStateOnMouseout);
```

Comment:
Instead of defining `updateTooltipStateOnMouseout` outside the `useEffect` can't you do this? This way we don't have to pass it as a dependency and argument types are inferred automatically

```js
    mapboxMap.on('mouseout', event => {
      setTooltipState(null);
    });
```